sunpy/map/mapbase.py

```diff
@@ -2501,6 +2502,68 @@ def _check_axes(self, axes, warn_different_wcs=False):
 
         return axes
 
+    def reproject_to(self, output_projection, algorithm='interp', **reproject_args):
+        """
+        Reproject the map
+
+        ..note::
+            This method requires the optional package `reproject` to be installed.
+
+        Parameters
+        ----------
+        output_projection : `dict` or `~astropy.wcs.WCS`
+            The destination FITS WCS header or WCS instance
+        algorithm : str
+            One of the supported `reproject` algorithms (see below)
+
+        Returns
+        -------
+        outmap : `~sunpy.map.GenericMap`
+            The reprojected map
+
+        Notes
+        -----
+        The supported `reproject` algorithms are:
+        * 'interp' for :func:`~reproject.reproject_interp`
+        * 'adaptive' for :func:`~reproject.reproject_adaptive`
+        * 'exact' for :func:`~reproject.reproject_exact`
+        See the respective documentation for these functions for additional keyword
+        arguments that are allowed.
+
+        If the output projection is specified by a `~astropy.wcs.WCS` instance, the
+        `shape_out` keyword argument must also be specified for the dimensions of the
+        output map.  The ordering of these dimensions is in NumPy order (i.e.,
+        secondary axis and then primary axis).
+        """
+        try:
+            import reproject
+        except ImportError:
+            raise ImportError("This method requires the optional package `reproject`.")
+
+        if not isinstance(output_projection, astropy.wcs.WCS):
+            output_projection = Header(output_projection)
```

Comment:
Why `Header` and not `WCS`?

Reply:
It's because the header can contain the pixel dimensions, but that information is not preserved in the `WCS` instance.

Reply:
It is. It is preserved as `WCS.pixel_shape` https://docs.astropy.org/en/stable/api/astropy.wcs.wcsapi.BaseLowLevelWCS.html#astropy.wcs.wcsapi.BaseLowLevelWCS.pixel_shape

Reply:
Ugh.  You are correct that it is available in the WCS instance, but that means `reproject` has a bug/feature in that it does not make use of `.pixel_shape` (or rather, `.array_shape` for the NumPy ordering of dimensions).  `reproject` shouldn't require that `shape_out` be specified if `.array_shape` exists.

Reply:
Okay, fixed/worked around.  @Cadair can explain whether this is a bug or a feature in `reproject`.

Reply:
That's probably a feature request for reproject if you fancy opening it there.